packages/sdk/wallets/wallet-hsm-azure/src/azure-key-vault-client.ts

```diff
@@ -13,6 +13,19 @@ import {
 } from '@celo/wallet-hsm'
 import { BigNumber } from 'bignumber.js'
 
+type SupportedCurve = 'P-256K' | 'SECP256K1'
+type SigningAlgorithm = 'ECDSA256' | 'ES256K'
+
+const SUPPORTED_CURVES: readonly SupportedCurve[] = ['P-256K', 'SECP256K1'] as const
+const SIGNING_ALGORITHM_FOR: Record<SupportedCurve, SigningAlgorithm> = {
+  'P-256K': 'ES256K',
+  SECP256K1: 'ECDSA256',
+}
+
+const isSupportedCurve = (curveName: string): curveName is SupportedCurve => {
```

Comment:
If `curveName` is not of type `SupportedCurve`, will this throw an exception?

Reply:
Not sure what you mean, might be a miss-understanding regarding how type guards like this work. Basically this function is meant to determine if `curveName` (any string) is a SupportedCurve a.k.a one of `'P-256K' | 'SECP256K1'`.

The function itself returns `true` or `false` but it also constrains the variable's type. So later when we do:
```typescript
    const curve = await this.getKeyCurve(keyName)
    // here curve is of type `string`
    if (!isSupportedCurve(curve)) {
      // here curve is still `string` because in this branch isSupportedCurve returns false
      throw new Error(`Key curve ${curve} is not supported, must be one of: ${SUPPORTED_CURVES}`)
    }
    // here curve is `SupportedCurve` because we throw if isSupportedCurve returns false, so it must be true
    // and that allows us to index in the SIGNING_ALGORITHM_FOR map which is a Record<SupportedCurve, SigningAlgorithm>
    const signingAlgorithm = SIGNING_ALGORITHM_FOR[curve]

```

Reply:
I guess my question is: If `getKeyCurve()` returns a string that is not a `SupportedCurve`, will the type guard return false or is it only a compile-time check and it will just return false due to the `indexOf(curveName) === -1`?

Reply:
So basically we're just telling the compiler that if that function evaluates to `true` it can restrict the type of `curveName` from `string` to `SupportedCurve`. The compiler can't directly figure that out from the function body. But the compiler can look at usages of this function and if it's part of an if-statement than it can use this information in each branch: `false` it's just a string, `true` it's a `SupportedCurve`. So that lets us make us of this extra information. But the function itself will never throw or anything like that. 

You should think of the syntax of `(curveName: string): curveName is SupportedCurve => {}` as if `curveName is SupportedCurve` is the return value of the function, so the function doesn't just return a boolean value, it answers the question `variable is Type`.

So if `getKeyCurve()` returns something that's not a SupportedCurve we get inside of the if statement above and throw that Error, if it then we continue executing the function and the compiler knows that `curve` is of the restricted type.

Reply:
Ah that's really cool. Thanks for the explanation!